samples/emoji-search/android-views/src/main/kotlin/com/example/redwood/emojisearch/android/views/EmojiSearchActivity.kt

```diff
@@ -92,26 +95,48 @@ class EmojiSearchActivity : ComponentActivity() {
       )
     }
 
-    treehouseLayout = TreehouseLayout(this, widgetSystem, onBackPressedDispatcher).apply {
-      treehouseContentSource.bindWhenReady(this, treehouseApp, codeListener)
+    val viewRoot = EmojiSearchViewRoot(context, scope)
+
+    treehouseLayout = TreehouseLayout(this, widgetSystem, onBackPressedDispatcher, viewRoot).apply {
+      treehouseContentSource.bindWhenReady(this, treehouseApp)
     }
     setContentView(treehouseLayout)
   }
 
-  private val codeListener: CodeListener = object : CodeListener() {
-    override fun onCodeDetached(
-      app: TreehouseApp<*>,
-      view: TreehouseView<*>,
-      exception: Throwable?,
+  private class EmojiSearchViewRoot(
+    context: Context,
+    private val scope: CoroutineScope,
+  ) : ViewRoot(context) {
+    private var restart: (() -> Unit)? = null
+
+    override fun contentState(
+      loadCount: Int,
+      attached: Boolean,
+      uncaughtException: Throwable?,
     ) {
-      if (exception != null) {
-        treehouseLayout.reset()
-        treehouseLayout.addView(
-          ExceptionView(treehouseLayout, exception),
-          LinearLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT),
-        )
+      if (uncaughtException != null) {
+        scope.launch {
+          delay(2_000.milliseconds)
```

Comment:
Not that it matters, but could be
```suggestion
          delay(2.seconds)
```